File: ocaml/parser.ml
let read_whole_file filename =
        let ch = open_in filename in
        let s = really_input_string ch (in_channel_length ch) in
        close_in ch;
        s

let print = Sys.argv.(1) = "-p"
let filename = Sys.argv.(2)

let () =
        let str = read_whole_file filename in
        match Json.parse str with
        | Error (e,l) ->
                Printf.printf "%s at byte %i\n" e l;
                exit 1
        | Ok json ->
                        if print then
                                Printf.printf "%a\n" Json.print json
                        else
                                Printf.printf "Ok\n"

File: cpp/parser.cpp
#include "json.hpp"
#include "matching.hpp"

#include <fstream>
#include <iostream>

std::string read_file(const char* file_name)
{
	std::ifstream f(file_name);
	std::string contents{std::istreambuf_iterator<char>(f), std::istreambuf_iterator<char>()};
	return contents;
}

int main(int argc, char *argv[])
{
	if (argc < 2) return 1;
	const bool print = argv[1] == std::string("-p");
	const char* file = argv[2];
	std::string json_str = read_file(file);
	match(parse(json_str),
		[&](Json json){
			if (print) std::cout << json << "\n";
			else std::cout << "OK\n";
		},
		[](Error e, std::string_view){
			std::cout << "Error " << e << "\n";
		});
	
	return 0;
}
